File: Node/architecture-patterns/src/controllers/order-controller.js
'use strict';
/**
 * src/controllers/order-controller.js — CONTROLLER LAYER (compliant)
 *
 * Architecture rules enforced here:
 *   ✓ Imports only from services — never from repositories or routes
 *   ✓ Translates between HTTP concerns and domain objects
 *   ✓ Delegates all business logic to order-service
 */

const service = require('../services/order-service');

function createOrder(req, res) {
  const { userId, amount } = req.body;
  const order = service.createOrder(userId, amount);
  res.status(201).json(order);
}

function getOrder(req, res) {
  const order = service.getOrder(req.params.orderId);
  if (!order) {
    return res.status(404).json({ error: 'not_found' });
  }
  return res.json(order);
}

function processOrder(req, res) {
  try {
    const result = service.processOrder(req.params.orderId);
    if (result.status === 'not_found') {
      return res.status(404).json(result);
    }
    return res.json(result);
  } catch (err) {
    return res.status(422).json({ error: err.message });
  }
}

module.exports = { createOrder, getOrder, processOrder };

File: dotnet/architecture-patterns/src/Controllers/OrderController.cs
using ArchitecturePatterns.Services;

namespace ArchitecturePatterns.Controllers;

/// <summary>
/// OrderController — CONTROLLER LAYER (compliant)
///
/// Architecture rules enforced here:
///   ✓ Uses only ArchitecturePatterns.Services — never Repositories
///   ✓ Translates request/response concerns to/from domain objects
///   ✓ Injects IOrderService via constructor
/// </summary>
public class OrderController
{
    private readonly IOrderService _service;

    public OrderController(IOrderService service)
    {
        _service = service;
    }

    public Dictionary<string, object> CreateOrder(string userId, decimal amount)
    {
        var order = _service.CreateOrder(userId, amount);
        return new Dictionary<string, object>
        {
            ["orderId"] = order.OrderId,
            ["status"]  = order.Status,
            ["amount"]  = order.Amount,
        };
    }

    public Dictionary<string, object> GetOrder(string orderId)
    {
        var order = _service.GetOrder(orderId);
        if (order is null)
            return new Dictionary<string, object> { ["error"] = "not_found" };

        return new Dictionary<string, object>
        {
            ["orderId"] = order.OrderId,
            ["status"]  = order.Status,
        };
    }

    public Dictionary<string, object> ProcessOrder(string orderId)
    {
        try
        {
            var order = _service.ProcessOrder(orderId);
            return new Dictionary<string, object>
            {
                ["orderId"] = order.OrderId,
                ["status"]  = order.Status,
            };
        }
        catch (ArgumentException ex)
        {
            return new Dictionary<string, object> { ["error"] = ex.Message };
        }
    }
}
